File: backend/app/schemas/duty_shift_handover.py
from datetime import date, datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field


class DutyShiftHandoverCreate(BaseModel):
    schedule_id: int = Field(..., description="ID dòng ca trực trong duty_schedules")
    giver_name: Optional[str] = Field(None, max_length=100, description="Tên người giao ca (mặc định lấy theo tài khoản)")
    receiver_id: Optional[int] = Field(None, description="ID tài khoản người nhận ca nếu có")
    receiver_name: Optional[str] = Field(None, max_length=100, description="Họ tên người nhận ca")
    personnel_report: Optional[str] = Field(None, description="Tình hình quân số: có mặt, vắng mặt có lý do...")
    equipment_status: Optional[str] = Field(None, description="Tình hình khí tài thông tin liên lạc, vũ khí trang bị")
    incident_log: Optional[str] = Field(None, description="Nhật ký các sự vụ, mệnh lệnh nhận được trong ca trực")
    pending_tasks: Optional[str] = Field(None, description="Nhiệm vụ còn dở dang bàn giao ca sau theo dõi, xử lý")


class DutyShiftHandoverAcknowledge(BaseModel):
    status: str = Field("da_nhan", pattern="^(da_nhan|co_kien_nghi)$", description="Trạng thái xác nhận: da_nhan hoặc co_kien_nghi")
    receiver_note: Optional[str] = Field(None, max_length=500, description="Ghi chú phản hồi hoặc kiến nghị của ca sau")


class DutyShiftHandoverCommanderReview(BaseModel):
    commander_note: str = Field(..., max_length=500, description="Ý kiến nhận xét, chỉ đạo của Chỉ huy ca trực")


class DutyShiftHandoverOut(BaseModel):
    id: int
    schedule_id: int
    duty_date: Optional[date] = None
    duty_type: Optional[str] = None
    shift: Optional[str] = None
    unit_name: Optional[str] = None
    giver_id: int
    giver_name: str
    receiver_id: Optional[int] = None
    receiver_name: Optional[str] = None
    handover_time: datetime
    personnel_report: Optional[str] = None
    equipment_status: Optional[str] = None
    incident_log: Optional[str] = None
    pending_tasks: Optional[str] = None
    commander_note: Optional[str] = None
    status: str
    receiver_note: Optional[str] = None
    acknowledged_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DutyShiftHandoverListResponse(BaseModel):
    items: list[DutyShiftHandoverOut]
    total: int
    skip: int
    limit: int

File: frontend/src/types/dutyShiftHandover.ts
// Dong bo openapi v8.0.0 - Bien ban ban giao ca truc & So nhat ky kip truc dien tu.
// v8.0.0 khong doi schema; chi siet hanh vi backend: acknowledge co guard trang thai +
// chi receiver duoc chi dinh moi ky (403/409), chan role 5 lap bien ban & tu ban giao (400/403).

export type DutyShiftHandoverStatus = 'cho_nhan' | 'da_nhan' | 'co_kien_nghi'

export const HANDOVER_STATUS_LABELS: Record<DutyShiftHandoverStatus, string> = {
  cho_nhan: 'Chờ nhận ca',
  da_nhan: 'Đã nhận ca',
  co_kien_nghi: 'Có kiến nghị',
}

export const HANDOVER_STATUS_COLORS: Record<
  DutyShiftHandoverStatus,
  { bg: string; text: string; border: string }
> = {
  cho_nhan: {
    bg: 'bg-amber-50 dark:bg-amber-950/40',
    text: 'text-amber-700 dark:text-amber-300',
    border: 'border-amber-200 dark:border-amber-800',
  },
  da_nhan: {
    bg: 'bg-emerald-50 dark:bg-emerald-950/40',
    text: 'text-emerald-700 dark:text-emerald-300',
    border: 'border-emerald-200 dark:border-emerald-800',
  },
  co_kien_nghi: {
    bg: 'bg-rose-50 dark:bg-rose-950/40',
    text: 'text-rose-700 dark:text-rose-300',
    border: 'border-rose-200 dark:border-rose-800',
  },
}

export interface DutyShiftHandover {
  id: number
  schedule_id: number
  duty_date: string | null
  duty_type: string | null
  shift: string | null
  unit_name: string | null
  giver_id: number
  giver_name: string
  receiver_id: number | null
  receiver_name: string | null
  handover_time: string
  personnel_report: string
  equipment_status: string
  incident_log: string | null
  pending_tasks: string | null
  commander_note: string | null
  status: DutyShiftHandoverStatus
  receiver_note: string | null
  acknowledged_at: string | null
  created_at: string
  updated_at: string
}

export interface DutyShiftHandoverCreate {
  schedule_id: number
  receiver_id?: number | null
  receiver_name?: string | null
  giver_name?: string | null
  personnel_report: string
  equipment_status: string
  incident_log?: string | null
  pending_tasks?: string | null
}

export interface DutyShiftHandoverAcknowledge {
  receiver_note?: string | null
  status?: 'da_nhan' | 'co_kien_nghi'
}

export interface DutyShiftHandoverCommanderReview {
  commander_note: string
}

export interface DutyShiftHandoverListResponse {
  items: DutyShiftHandover[]
  total: number
}
